File: Model/DotNetComponentInterface/ApsimAttributes.cpp
using namespace System;

[AttributeUsage(AttributeTargets::Method)]
public ref class EventHandler : public System::Attribute
	{};
	
[AttributeUsage(AttributeTargets::Event)]
public ref class Event : public System::Attribute
	{};
	

[AttributeUsage(AttributeTargets::Field | AttributeTargets::Property, AllowMultiple = true)]
public ref class Param : public System::Attribute
	{
	public:
		String^ Name;
		bool Optional;
		double MinVal;
		double MaxVal;
		Param()
			{
			Name = "";
			Optional = false;
			MinVal = MaxVal = Double::NaN;  // Using Nan to indicate min. and max. values not set
			}
		Param(bool IsOptional)
			{
			Name = "";
			Optional = IsOptional;
			MinVal = MaxVal = Double::NaN;  // Using Nan to indicate min. and max. values not set
			}
		Param(String^ Name)
			{
			this->Name = Name;
			Optional = false;
			MinVal = MaxVal = Double::NaN;
			}
		Param(String^ Name, bool IsOptional)
			{
			this->Name = Name;
			Optional = IsOptional;
			MinVal = MaxVal = Double::NaN;
			}
 	};

[AttributeUsage(AttributeTargets::Field | AttributeTargets::Property, AllowMultiple = true)]
public ref class Input : public System::Attribute
	{
	public:
	   bool Optional;
	   Input()
	      {
	      Optional = false;
	      }
	   Input(bool IsOptional)
	      {
	      Optional = IsOptional;
	      }
	};
	
[AttributeUsage(AttributeTargets::Field | AttributeTargets::Property, AllowMultiple = true)]
public ref class Output : public System::Attribute
	{
	public:
		String^ Name;
		Output()
			{
			Name = "";
			}
		Output(String^ Name)
			{
			this->Name = Name;
			}	   
	};
[AttributeUsage(AttributeTargets::Field, AllowMultiple = true)]
public ref class Writable : public System::Attribute
	{
	};	
[AttributeUsage(AttributeTargets::Field  | AttributeTargets::Property, AllowMultiple = true)]
public ref class Units : public System::Attribute
	{
	private:
		String^ St;
		
	public:
		Units(String^ Units) 
			{ 
			St = Units;
			}

		virtual String^ ToString() override 
			{
			return St;
			}

	};	
[AttributeUsage(AttributeTargets::All, AllowMultiple = true)]
public ref class Description : public System::Attribute
	{
	private:
		String^ St;
		
	public:
		Description(String^ Description) 
			{ 
			St = Description;
			}

		virtual String^ ToString() override 
			{
			return St;
			}

	};	
	

[AttributeUsage(AttributeTargets::Class, AllowMultiple = true)]
public ref class Model : public System::Attribute
	{
	};	

public enum class IsOptional {Yes, No};

[AttributeUsage(AttributeTargets::Field, AllowMultiple = true)]
public ref class Link : public System::Attribute
	{
   public:
      String^ _Path;
      IsOptional _IsOptional;
      Link() {_Path = nullptr; _IsOptional = IsOptional::No;}
      Link(IsOptional isOptional) {_Path = nullptr; _IsOptional = isOptional;}
      Link(String^ Path) {_Path = Path; _IsOptional = IsOptional::No;}
      Link(String^ Path, IsOptional isOptional) {_Path = Path; _IsOptional = isOptional;}
	};	

File: Model/CSDotNetComponentInterface/ApsimAttributes.cs
using System;

/// <summary>
/// These classes define the attributes used to provide metadata for the 
/// APSIM Component properties and events.
/// </summary>

[AttributeUsage(AttributeTargets.Method)]
public class EventHandler : Attribute
{
    private string _EventName;

    public EventHandler()
    {
        _EventName = "";
    }

    public EventHandler(String eventName)
    {
        _EventName = eventName;
    }

    public string EventName
    {
        get { return _EventName; }
        set { _EventName = EventName; }
    }
}

[AttributeUsage(AttributeTargets.Event)]
public class Event : Attribute
{
}


[AttributeUsage(AttributeTargets.Field | AttributeTargets.Property, AllowMultiple = true)]
public class Param : Attribute
{
    private string _Name = "";
    private bool _Optional = false;
    private double _MinVal = Double.NaN;
    private double _MaxVal = Double.NaN;

    public string Name
    {
        get { return _Name; }
        set { _Name = value; }
    }

    public bool IsOptional
    {
        get { return _Optional; }
        set { _Optional = value; }
    }

    public double MinVal
    {
        get { return _MinVal; }
        set { _MinVal = value; }
    }

    public double MaxVal
    {
        get { return _MaxVal; }
        set { _MaxVal = value; }
    }

}

[AttributeUsage(AttributeTargets.Field | AttributeTargets.Property, AllowMultiple = true)]
public class Input : Attribute
{
    private bool Optional = false;

    public bool IsOptional
    {
        get { return Optional; }
        set { Optional = value; }
    }
}

[AttributeUsage(AttributeTargets.Field | AttributeTargets.Property, AllowMultiple = true)]
public class Output : Attribute
{
    public String Name;
    public Output()
    {
        Name = "";
    }
    public Output(String sName)
    {
        Name = sName;
    }
}

[AttributeUsage(AttributeTargets.Field, AllowMultiple = true)]
public class Writable : Attribute
{
}

[AttributeUsage(AttributeTargets.Field | AttributeTargets.Property, AllowMultiple = true)]
public class Units : Attribute
{
    private String St;
    public Units(String Units)
    {
        St = Units;
    }
    public override String ToString()
    {
        return St;
    }
}

[AttributeUsage(AttributeTargets.All, AllowMultiple = true)]
public class Description : Attribute
{
    private String St;

    public Description(String Description)
    {
        St = Description;
    }

    public override String ToString()
    {
        return St;
    }
}


[AttributeUsage(AttributeTargets.Class, AllowMultiple = true)]
public class Model : Attribute
{
}

[AttributeUsage(AttributeTargets.Field, AllowMultiple = true)]
public class Link : Attribute
{
    private String _Path = null;
    private bool _IsOptional = false;

    //public Link() { }
    //public Link(string NamePath, bool IsOptional) { _Path = NamePath; _IsOptional = IsOptional; }
    public string NamePath
    {
        get { return _Path; }
        set { _Path = value; }
    }
    public bool IsOptional
    {
        get { return _IsOptional; }
        set { _IsOptional = value; }
    }
   
}


/// <summary>
/// This attribute is used in DotNetProxies.cs to provide a map between the 
/// proxy class and the component type.
/// Usage: [ComponentType("Plant2")]
/// </summary>
[AttributeUsage(AttributeTargets.Class, AllowMultiple = false)]
public class ComponentTypeAttribute : Attribute
{
    public String ComponentClass;
    public ComponentTypeAttribute(String CompClass)
    {
        ComponentClass = CompClass;
    }

}

